python/phase-2.md

```diff
@@ -0,0 +1,830 @@
+        while True:
+            # Store greeting (this is typed)
+            self._current_greeting = await workflow.execute_local_activity(
+                create_greeting_activity,
+                self._salutation,
+                name,
+                start_to_close_timeout=timedelta(seconds=5),
+            )
+            workflow.logger.debug("Greeting set to %s", self._current_greeting)
+            
+            # Wait for salutation update or complete signal (this can be
+            # cancelled)
+            await asyncio.wait(
+                [self._salutation_update.wait(), self._complete.wait()],
+                return_when=asyncio.FIRST_COMPLETED,
+            )
+            if self._salutation_update.is_set():
+                self._salutation_update.clear()
+            elif self._complete.is_set():
+                return self._current_greeting
+
+    @workflow.signal
+    async def update_salutation(self, salutation: str) -> None:
+        self._salutation = salutation
+        self._salutation_update.set()
+
+    @workflow.signal
+    async def complete_with_greeting(self) -> None:
+        self._complete.set()
+
+    @workflow.query
+    async def current_greeting(self) -> str:
+        return self._current_greeting
+
+def test_workflow(client: Client) -> None:
+    # Start the workflow
+    handle = await client.start_workflow(GreetingWorkflow.run, "John")
+
+    # Confirm the current greeting
+    assert "Hello, John!" == await handle.query(GreetingWorkflow.current_greeting)
+
+    # Change the salutation part
+    await handle.signal(GreetingWorkflow.update_salutation, "Hola")
+
+    # Now tell it to complete with the new greeting
+    await handle.signal(GreetingWorkflow.complete_with_greeting)
+
+    # Check result
+    assert "Hola, John!" == await handle.result()
+
+def create_worker(client: Client) -> Worker:
+    return Worker(
+      client=client,
+      activities=[create_greeting_activity],
+      workflows=[GreetingWorkflow],
+    )
+```
+
+Notes:
+
+* Workflows are classes
+  * Consideration was given to making workflows be functions (in addition to classes or having the class equivalent sit
+    on top) with several possibilities for defining signals and queries, but all were too ugly/complicated
+* `@activity.defn`
+  * Must be on function and accepts an optional activity `name` which defaults to function `__name__`
+  * This was originally not wanted in the activity impl (phase 1) due to complaints of the name not being obvious
+  * We need to define the activity name on the activity for execution by the workflow
+  * Is it ok to now make this required? Is it ok to now make the name implicit? Is it ok to now use a list instead of a
+    dict to specify activities for the worker?
+* `@workflow.defn`
+  * Must be on class and accepts an optional workflow `name` which defaults to class `__name__`
+  * This is required to clearly mark the class as a workflow
+  * No base classes can have this value
```

Comment:
Can you elaborate here? Can I not extend a workflow class? Why?

Reply:
The primary reason is to avoid the "Diamond Problem" when resolving attribute ambiguity. I think it needs to be clear that the workflow is sealed/final so the reader is not surprised by the actual implementation.

Having said that, with how I mentioned abstract classes/protocols could be used, I think we can relax this constraint to say that "Every registered workflow class must have the decorator" so even if you extend one, you must put the decorator on yourself since it's essentially a marker for name.

Reply:
I have updated this to allow base classes to have the decorator but it is not inherited